content/2018/12/10/channels-and-drf.rst

```diff
@@ -0,0 +1,292 @@
+with parallel structure to the code. This makes it easier to exclude
+them on an install.
+
+Inside those files, we mostly have simple declarative use of DRF. Follow
+their `tutorial`_ if you want to get that set up. We run our tests with
+`pytest`_ and a requirement of 100% test coverage, as we do with all our
+projects.
+
+Channels is perhaps a little less likely to be familiar; I haven't seen
+it get quite as much uptake, being a more specialized tool. That said,
+it's pretty simple to set up on top of an existing Django application,
+so I'll walk you through it in brief.
+
+First, you need to move from using WSGI to ASGI, which is "basically the
+same, but async". This means changing your server process from gunicorn
+to something like Daphne, changing your ``project.wsgi`` file to
+``project.asgi`` (as described in the `Channels docs`_), add a
+``routing`` module and a ``consumers`` module, and adjust your settings
+appropriately.
+
+At this stage, you won't yet have anything in ``consumers``, and you
+won't have much in routing. It can look like this::
+
+
+   from channels.routing import ProtocolTypeRouter
+
+   application = ProtocolTypeRouter({
+       # (http->django views is added by default)
+   })
+
+Yep, that's basically an empty ``ProtocolTypeRouter``. We're just making
+sure we don't break anything with the transition to ASGI, first, and
+``ProtocolTypeRouter`` does as the comment says, and wires HTTP to
+Django by default.
+
+Once that's all done and you've confirmed that everything's still
+working, you can start to add in the wiring for WebSockets.
+
+WebSockets, Consumers, and Groups
+---------------------------------
+
+Let's talk a bit about architecture before we talk about how to
+implement it.
+
+As we're using it, Channels primarily drives the pushing of
+notifications about the state of selected objects to the client, over
+WebSockets. We've opted to simplify the client's job by having one
+endpoint that it can call to subscribe to any objects it wants, and to
+use the payload it sends to validate and set up that subscription. So
+the client sends the following data to
+``wss://server.domain/ws/notifications``::
+
+   {
+       "object_type": "string representation of type",
+       "id": "hashid of the instance"
+   }
+
+The server will then decide if the requesting user can subscribe to that
+model, and start sending them updates over that WebSocket if so.
+
+On the server's side of things, we have a Consumer object that handles a
+bunch of WebSocket events, and, when appropriate, adds a particular
+socket connection to a named Group. Elsewhere in the server logic, we
+send events to that group when the model changes, and all subscribed
+sockets will receive a serialization of the model with the changes.
+
+(Since we're using React on the frontend for this project, we're also
+sending a value that happens to map to the Redux event names we're
+using, but that sort of tight coupling may not match your needs.)
```

Comment:
This is partially accurate -- we're not 100% consistent between WS event names and Redux action types. But I think the point stands, and it's fine as-is.